Range check via Cprover assert

xbauch opened the pull request.

hannes-steffenhagen-diffblue commented on gnat2goto/driver/tree_walk.adb:
```diff
@@ -377,6 +390,19 @@ package body Tree_Walk is
+   function Uint_To_Integer (Number : Uint) return Integer
+   is
+   begin
+      return Integer (UI_To_Int (Number));
```
This restricts us to 32 bit bounds (`Int` is basically a 32 bit signed integer). We need to find a better way around this

xbauch replied:
@hannes-steffenhagen-diffblue I've added one commit that addresses that.

xbauch replied:
The commit is gone now, but basically we went from Int to Uint (which is of arbitrary size). It is now in separate package `range_check`.

hannes-steffenhagen-diffblue commented on gnat2goto/driver/tree_walk.adb:
```diff
@@ -113,13 +115,20 @@ package body Tree_Walk is
+   function Do_Assert (N : Node_Id; Assertion : Irep; Description : Irep)
+                       return Irep;
+
+   function Do_Range_Assert (N : Node_Id; Value : Irep; Value_Type : Irep)
```
These 3 functions actually generate code, but they're using the naming scheme of the functions that handle an AST node. I think we need to come up with less confusing names for these.

hannes-steffenhagen-diffblue commented on gnat2goto/driver/tree_walk.adb:
```diff
@@ -928,8 +1029,10 @@ package body Tree_Walk is
          Set_Width (R, Get_Width (Resolved_Underlying));
-         Set_Lower_Bound (R, Do_Constant (Low_Bound (Range_Expr)));
-         Set_Upper_Bound (R, Do_Constant (High_Bound (Range_Expr)));
+         Set_Lower_Bound (R, Uint_To_Integer (Intval (Low_Bound
```
I still don't think it makes much sense for these to be properties on the type; For dynamic bounds at least we'll need to carry around the bounds in variables/hidden parameters anyway

hannes-steffenhagen-diffblue replied:
What is keeping us from using unbounded signedbv types for the bounds here by the way?

xbauch replied:
@hannes-steffenhagen-diffblue I'm not sure what is going to be the best approach for dynamic bounds. As of now, the bounds are `Uint`s (as per the last commit) and the `Irep` properties are `Integer`. Having any other numeric type inside `Irep`s would require changing the schemes, and the generating Python script.

xbauch commented:
@hannes-steffenhagen-diffblue I've renamed the 3 functions that create code (to `Make_*_Expr/Call`).  Changed the expected output of `ranges` test (which is intended as a test for this feature). And generalised the bounds to `Uint`s which are stored in a separate hash-table and only indexed from `Irep`s for bounded types. I squashed it to 1 commit, but can break it again.

hannes-steffenhagen-diffblue commented on gnat2goto/driver/tree_walk.adb:
```diff
@@ -1483,12 +1632,24 @@ package body Tree_Walk is
+                      Value => Sym_Type);
+            Set_Value (I     => Lower_Bound,
+                       Value => Load_Bound_In_Hex (Get_Lower_Bound (Sym_Type),
+                        Pos (Get_Width (Sym_Type))));
```
What if the width of the type isn't large enough to hold the value?

xbauch replied:
Wouldn't that mean that the bound cannot be stored in a variable of that type?

hannes-steffenhagen-diffblue commented on testsuite/gnat2goto/tests/ranges/test.out:
```diff
@@ -1 +1,3 @@
-VERIFICATION SUCCESSFUL
+[assertion.1] Range Check: FAILURE
+[assertion.2] Range Check: FAILURE
```
It'd probably be better to have a test where both failure and successes happen, and that's also testing some things like casting into a larger or smaller type

xbauch replied:
I've extended the existing test with a few more cases, making the range check pass as well. I'm not sure it cover all reasonable cases.

hannes-steffenhagen-diffblue commented on gnat2goto/ireps/irep_specs/bounded_signedbv_type.json:
```diff
@@ -1,7 +1,7 @@
-    "lower_bound": {"schema": "constant_expr"},
-    "upper_bound": {"schema": "constant_expr"}
+    "lower_bound": {"type": "integer"},
+     "upper_bound": {"type": "integer"}
```
stray space

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on testsuite/gnat2goto/tests/ranges/ranges.adb:
```diff
@@ -3,9 +3,13 @@ procedure Ranges is
    type Int2 is new Int1 range 2 .. 9;
    subtype Int3 is Int1 range 3 .. 8;
```
The test should make sure that the integer types here genuinely have different sizes (in bits), not just different ranges

xbauch commented on gnat2goto/driver/range_check.ads:
```diff
@@ -0,0 +1,32 @@
+                                 Hash            => Uint_Hash,
+                                 Equivalent_Keys => "=");
+
+   Bounds_Integer_Table : Bounds_Integer_Maps.Map;
```
Yes and no: `Integer_Bounds_Table` is used in a prereq. Moved.

xbauch commented on gnat2goto/driver/tree_walk.adb:
```diff
@@ -880,6 +893,90 @@ package body Tree_Walk is
+
+      -----------------------------
+      -- Build_Assert_Function --
+      -----------------------------
```
I didn't feel like changing the scheme set up in `tree_walk`.

xbauch commented on gnat2goto/driver/tree_walk.adb:
```diff
@@ -880,6 +893,90 @@ package body Tree_Walk is
+
+      function Build_Assert_Function return Symbol
+      is
+         Func_Type : constant Irep := New_Irep (I_Code_Type);
```
Done.

xbauch replied:
Undone, only `Class_Expr` ireps can have source location.

xbauch commented on gnat2goto/driver/tree_walk.adb:
```diff
@@ -892,23 +989,31 @@ package body Tree_Walk is
+
+      declare
+         R : constant Irep := New_Irep (I_Code_Assign);
+      begin
```
The `return` was getting uncomfortably long, other than that I also think they are equivalent.

xbauch commented on gnat2goto/driver/tree_walk.adb:
```diff
@@ -1199,9 +1308,14 @@ package body Tree_Walk is
+                    Convert_Uint_To_Hex
                       (Intval (N), Pos (Constant_Width)));
+      else
+         Set_Value (Ret, UI_Image (Input  => Intval (N),
```
I'm not sure, I might have introduced it because the ranges were integer literals and this function did not handle them correctly. Maybe.

sonodtt replied:
@xbauch this really needs some comments. 
This change was prompted by something - please tell us the reason behind the 'else'.
If this is one of those 'this works' kind of things that is hazy - please tell us what didn't work.

xbauch replied:
@sonodtt I will find out and describe it in the code. It will take a bit longer though.

xbauch commented on gnat2goto/driver/range_check.adb:
```diff
@@ -0,0 +1,49 @@
+
+   function Uint_Hash (Id : Bound_Type) return Ada.Containers.Hash_Type is
+   begin
+      return Ada.Containers.Hash_Type (UI_To_Int (Id));
```
Simplified to a vector with potential duplicates.

hannes-steffenhagen-diffblue commented on gnat2goto/driver/tree_walk.adb:
```diff
@@ -880,6 +894,97 @@ package body Tree_Walk is
+      New_Symbol.Mode := Intern ("C");
+      New_Symbol.Value := Value;
+
+      if not (Global_Symbol_Table.Contains (Key => Intern (Name))) then
```
The `else` keys in this instance seems like it'd be an error

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on gnat2goto/driver/tree_walk.adb:
```diff
@@ -21,13 +21,18 @@ with Ada.Exceptions;
    procedure Append_Declare_And_Init
      (Symbol : Irep; Value : Irep; Block : Irep; Source_Loc : Source_Ptr);
 
+   function Add_To_Table (Name : String; Symbol_Type : Irep; Value : Irep)
```
Could use a slightly clearer name. Maybe `New_Symbol_Table_Entry` or something like that?

hannes-steffenhagen-diffblue commented on gnat2goto/driver/tree_walk.adb:
```diff
@@ -880,6 +894,97 @@ package body Tree_Walk is
+         Func_Args : constant Irep := New_Irep (I_Parameter_List);
+         Func_Name : constant String :=
+           Fresh_Var_Name ("range_check");
+         Body_Block : constant Irep := New_Irep (I_Code_Block);
```
:pick: I'd prefer `Make_XXX` functions in place of these

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on gnat2goto/driver/tree_walk.adb:
```diff
@@ -880,6 +894,97 @@ package body Tree_Walk is
+   function Make_Range_Assert_Expr (N : Node_Id; Value : Irep;
+                                                 Value_Type : Irep)
+                                    return Irep
+   is
```
This could use a comment showing (in pseudocode) what kind of code is being generated here

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on gnat2goto/driver/tree_walk.adb:
```diff
@@ -1575,25 +1788,12 @@ package body Tree_Walk is
               Fresh_Var_Symbol_Expr (Type_Irep, Func_Str);
             SE_Call_Expr : constant Irep :=
-              New_Irep (I_Side_Effect_Expr_Function_Call);
+              Make_Assume_Expr (N, Make_Range_Expression (Sym_Nondet,
```
:pick: It might make sense to put the definition of `Make_Assume_Expr` and `Make_Assert_Call` in a separate commit, although I'd see that as pretty optional

xbauch commented on gnat2goto/driver/goto_utils.ads:
```diff
@@ -29,6 +29,15 @@ package GOTO_Utils is
    function Symbol_Expr (Sym : Symbol) return Irep
    with Post => Kind (Symbol_Expr'Result) = I_Symbol_Expr;
 
+   function Create_Fun_Parameter (Fun_Name : String; Param_Name : String;
```
Done.

xbauch commented on gnat2goto/driver/goto_utils.adb:
```diff
@@ -92,6 +92,42 @@ package body GOTO_Utils is
+                             Identifier      => Unintern (Func_Param_Id));
+   begin
+      A_Symbol_Table.Insert (Func_Param_Id, Value_Symbol);
+      Set_Type (Value_Arg, Param_Type);
```
Done.

xbauch commented on gnat2goto/driver/goto_utils.adb:
```diff
@@ -92,6 +92,42 @@ package body GOTO_Utils is
+      A_Symbol_Table.Insert (Func_Param_Id, Value_Symbol);
+      Set_Type (Value_Arg, Param_Type);
+      Set_Identifier (Value_Arg, Unintern (Func_Param_Id));
+      Set_Base_Name (Value_Arg, "value");
```
Done.

sonodtt commented on gnat2goto/driver/tree_walk.adb:
```diff
@@ -1546,7 +1710,56 @@ package body Tree_Walk is
+   function Make_Assert_Call (N : Node_Id; Assertion : Irep;
+                                           Description : Irep)
+                             return Irep is
+      SE_Call_Expr : constant Irep :=
```
'SE_Call_Expr' - in particular the SE.  This is no longer a side effect call.

xbauch commented on gnat2goto/driver/tree_walk.adb:
```diff
@@ -880,6 +894,97 @@ package body Tree_Walk is
+
+         Set_Parameters (Func_Type, Func_Args);
+         Set_Return_Type (Func_Type, Actual_Type);
+         Set_Ellipsis (Func_Type, False);
```
It was among the things we thought might fix the parameter-assignment problem. It wasn't, but I think it denotes the use of parameter packs. There should be none, hence false.
